Call init's parameter interaction before we create the UI options model

jonasschnelli opened the pull request:
Currently `optionsModel`(QT) parameters interaction overwrites/dominates initial and depending parameter settings. This PR would factor out init's parameter interaction and call it before `AppInit2()`.

Fixes #6773

jonasschnelli committed:
Refactor parameter interaction, call it before AppInit2()

jonasschnelli committed:
[QT] Call inits parameter interaction before we create the options model

jonasschnelli committed:
Move -blocksonly parameter interaction to the new ParameterInteraction() function

jonasschnelli commented:
Force pushed a copy'n'paste fix: `s/bool InitParameterInteraction()/void InitParameterInteraction()`.
Credits to @MarcoFalke on IRC.

MarcoFalke commented:
Concept ACK.

laanwj commented:
Concept ACK

paveljanik commented:
ACK

laanwj commented on src/qt/bitcoin.cpp:
```diff
@@ -381,7 +398,7 @@ void BitcoinApplication::startThread()
     coreThread = new QThread(this);
-    BitcoinCore *executor = new BitcoinCore();
+    executor = new BitcoinCore();
     executor->moveToThread(coreThread);
```
It's not allowed to use this object after moving it to another thread.
I'd suggest calling `InitParameterInteraction()` directly from `BitcoinApplication::parameterSetup()` and not bothering with the stuff in BitcoinCore.

jonasschnelli commented:
Rebased and fixed the pointer access after `moveToThread` by following @laanwj advice.

gmaxwell commented:
Sorry, blocksonly mode caused a need for a small change here, as it needs to get moved. @jonasschnelli  mind rebasing?

What about that prune/txindex "interaction" (fails to run), should it move too?

jonasschnelli commented:
> Sorry, blocksonly mode caused a need for a small change here, as it needs to get moved. @jonasschnelli mind rebasing?

Just rebased and moved the `-blocksonly` interaction to the new function.

> What about that prune/txindex "interaction" (fails to run), should it move too?

My idea was to only move "rules" to the new function. Things that evaluate parameters through `GetBoolArg` or `mapArgs`, etc. and depending on its state, optional change another startup variable through `SoftSetBoolArg`, etc.

Pruning and TxIndex are checks that lead to a `InitError()` (startup termination) and it would require to move the `InitError()` context to the `ParameterInteraction()` function.

MarcoFalke commented:
utACK 51627cc

jonasschnelli commented:
Would be nice if someone finds time to test this. Binaries: https://bitcoin.jonasschnelli.ch/pulls/6780/
This fixes a ugly bug (#6773) which would be nice to have fixed in 0.12.

fanquake commented:
Will do.

On Saturday, November 21, 2015, Jonas Schnelli <notifications@github.com>
wrote:

> Would be nice if someone finds time to test this. Binaries:
> https://bitcoin.jonasschnelli.ch/pulls/6780/
> This fixes a ugly bug (#6773
> <https://github.com/bitcoin/bitcoin/issues/6773>) which would be nice to
> have fixed in 0.12.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/6780#issuecomment-158617257>.
>

MarcoFalke commented:
@jonasschnelli 

* master: 

```
$ src/bitcoind -regtest -connect=127.0.0.1 -server -printtoconsole |grep parameter 
2015-11-21 17:52:56 AppInit2: parameter interaction: -connect set -> setting -dnsseed=0
2015-11-21 17:52:56 AppInit2: parameter interaction: -connect set -> setting -listen=0
2015-11-21 17:52:56 AppInit2: parameter interaction: -listen=0 -> setting -upnp=0
2015-11-21 17:52:56 AppInit2: parameter interaction: -listen=0 -> setting -discover=0
2015-11-21 17:52:56 AppInit2: parameter interaction: -listen=0 -> setting -listenonion=0

$ src/qt/bitcoin-qt -regtest -connect=127.0.0.1 -server -printtoconsole |grep parameter 
2015-11-21 17:53:34 AppInit2: parameter interaction: -connect set -> setting -dnsseed=0
```

* 51627cc

```
$ ~/bitcoin* -regtest -connect=127.0.0.1 -debug=1 -printtoconsole |grep parameter 
```

dcousens commented:
concept ACK / utACK

jonasschnelli commented:
I'd like to take this into 0.12 because it fixes a slightly security-critical issue (#6773). Some tested ACKs would be nice (another 5 days left until feature freeze).

MarcoFalke commented:
@jonasschnelli I am happy to test but the debug.log issue seems not yet resolved? Am I missing something?

laanwj commented:
utACK. Agree it needs to be in 0.12

Haven't tested any specific cases - I've been holding off on this because the parameter processing is kind of fragile and I haven't properly reasoned it through yet that this doesn't have effect on the precedence of options specified in e.g. configuration files, command line, QSettings.

jonasschnelli committed:
Initialize logging before we do parameter interaction

jonasschnelli commented:
Thanks @MarcoFalke for pointing out the logging issue. Added another commit that moves the logging initialization.

For reviewers: The critical part are the first 100 lines in `AppInit2()` because the logging init and the parameter interactions are now before `AppInit2()`.

MarcoFalke commented:
Sanity tests pass now:

```
$ src/bitcoind -regtest -connect=127.0.0.1 -server -printtoconsole |egrep "(parameter|Bitcoin)"
2015-11-27 07:39:08 Bitcoin version v0.11.99.0-296c42e (2015-11-26 14:03:27 +0100)
2015-11-27 07:39:08 InitParameterInteraction: parameter interaction: -connect set -> setting -dnsseed=0
2015-11-27 07:39:08 InitParameterInteraction: parameter interaction: -connect set -> setting -listen=0
2015-11-27 07:39:08 InitParameterInteraction: parameter interaction: -listen=0 -> setting -upnp=0
2015-11-27 07:39:08 InitParameterInteraction: parameter interaction: -listen=0 -> setting -discover=0
2015-11-27 07:39:08 InitParameterInteraction: parameter interaction: -listen=0 -> setting -listenonion=0

$ src/qt/bitcoin-qt -regtest -connect=127.0.0.1 -server -printtoconsole |egrep "(parameter|Bitcoin)"
2015-11-27 07:53:02 Bitcoin version v0.11.99.0-296c42e (2015-11-26 14:03:27 +0100)
2015-11-27 07:53:02 InitParameterInteraction: parameter interaction: -connect set -> setting -dnsseed=0
2015-11-27 07:53:02 InitParameterInteraction: parameter interaction: -connect set -> setting -listen=0
2015-11-27 07:53:02 InitParameterInteraction: parameter interaction: -listen=0 -> setting -upnp=0
2015-11-27 07:53:02 InitParameterInteraction: parameter interaction: -listen=0 -> setting -discover=0
2015-11-27 07:53:02 InitParameterInteraction: parameter interaction: -listen=0 -> setting -listenonion=0

MarcoFalke commented:
utACK 296c42e

MarcoFalke commented on src/init.cpp:
```diff
@@ -821,16 +839,6 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
-            LogPrintf("%s: parameter interaction: -blocksonly=1 -> setting -walletbroadcast=0\n", __func__);
-#endif
-    }
-    
```
@jonasschnelli Trivial white space rebase conflict. (Looks like you can ignore changes made in master)

jonasschnelli commented:
Thanks @MarcoFalke: rebased!
